src/net_processing.cpp

```diff
@@ -2822,6 +2822,49 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         return;
     }
 
+    // Received from a peer demonstrating readiness to announce transactions via reconciliations.
+    // This feature negotiation should happen:
+    // - between VERSION and VERACK to avoid relay problems from switching announcement protocols
+    //   after the connection is up
+    // - after WTXID because we reconcile only with peers supporting wtxid-relay
+    if (msg_type == NetMsgType::SENDRECON) {
+        if (pfrom.fSuccessfullyConnected) {
+            // Disconnect peers that send a SENDRECON message after VERACK.
+            LogPrint(BCLog::NET, "sendrecon received after verack from peer=%d; disconnecting\n", pfrom.GetId());
+            pfrom.fDisconnect = true;
+            return;
+        }
+
+        if (!pfrom.m_tx_relay) {
```

Comment:
I think you are right. This case shouldn't get us disconnected from the peer.

What would be the best way to detect peers which clearly do something meaningless?
I mean, this check could be just dropped I guess, but ideally, it should handle misbehavior.

Reply:
I don't really think of this as misbehaviour: I think of the meaning of SENDRECON as "I understand reconciliation and would like to participate in it on this connection" - if just on of the two sides signal that - ok, no reconciliation on this connection, but no harm done. So I think it might make sense to drop the check.

That being said, I think it might still make sense to not send SENDRECON when our peer has signified `fRelay=false` in their version message. Obviously, this would also mean no reconciliation with BIP37 peers, but that seems also like a good thing?! 
This has the additional benefit that no reconciliations would happen when one side is in blocksonly mode (i.e. has `m_tx_relay!=nullptr` but still doesn't wish to receive transactions). I think that currently, both sides would send SENDRECON in this case. One could also prevent this on the other side, by not sending out SENDRECON ourselves when in `-blocksonly` mode (`m_ignore_incoming_txs` set)

Reply:
I agree with most of what you say except the BIP37 part. Are you implying BIP37 peers won't want reconciliations?

Reply:
Not sure. I don't know too much about BIP37 and its usage, but I thought it unlikely that SPVs using it would go through the efforts of implementing it and get significant gains from it when they are interested in a limited number of transactions anyway. Do you think participating in recons would make sense for them?

The issue is that BIP37 peers initially set `fRelay=false` in their version and later activate tx relay by sending a filter. So, if BIP37 peers should use reconciliation, disabling SENDRECON based on receiving `fRelay=false` is no longer possible, and all the logic needs to be on the side of the peer that sends `fRelay=false` (although this work just fine).

Reply:
First of all, I'd note BIP37 user != SPV. You can think of another kind of implementation of BIP37, which filters, for example, 25% of all the addr space (not sure it's practical currently w.r.t filter size), doing that over N=20 connections with different filters to receive all transactions overall.
Given how unrealistic that is today, we can probably implement Erlay support for that once we know about this scenario?

Secondly, let's consider a regular small-filter (say, 1000 of addrs for redundancy) SPV node. If it's connected to X serving nodes, it'd still get X times more tx announcements than needed. But considering how little is that bandwidth, reducing this redundancy might be useless, so yeah.

At the same time, I don't think this justification is particularly strong, but yeah, in the end we could just suggest SPV nodes to not send SENDRECON if they don't want. In that case, the outcome is that regular nodes would send `SENDRECON` even though 'fRelay=false', and in some cases that would be useless, but that's not a big deal. 

Do you agree with this thinking?

Reply:
Yes, I agree it makes sense. To sum up, the suggestion would be 
1) to remove the `pfrom.m_tx_relay` check when receiving SENDRECON in order to not disconnect all block-relay-only peers
2) to add a check for `m_ignore_incoming_txs`  [here](https://github.com/naumenkogs/bitcoin/blob/ea3b87af90e1649428cabac83012d6faa8df61f5/src/net_processing.cpp#L2586) in order to opt out of recons when in `-blocksonly` mode

Reply:
Catching up to this discussion (sorry for missing it earlier). I'm not convinced that we care enough about doing tx reconciliation with bloom filtered peers to justify creating reconciliation-related state for them. If a peer is filtering transactions, obviously you don't care reconciling mempools: they clearly don't want all of your mempool's transactions, and most likely you won't learn much from them. Bloom filters are also disabled by default.